Unclear directions for PAT entering during `alr publish`

**Checklist**

- [x] I've searched for similar issues and found none.
- [x] I've included the output of `alr version`.
- [ ] I've included complete steps to reproduce my issue.

**Describe the bug**
`alr publish` fails in step 10 of 11:
```sh
fatal: unable to access 'https://github.com/jquorning/alire-index/': URL using bad/illegal format or missing URL
error: Could not complete the publishing assistant:
error:    Command ["git", "push", "-f", "-u", "writable", "release/hostarm-25.0.1"] exited with code 128
```

**To Reproduce**
Steps to reproduce the behavior:
1. Write a project
2. Run `alr publish`
3. See error

**Expected Behavior**
Expected alr publish to publish the crate to alire-index.

**`alr` version**
```sh
APPLICATION
alr version:               3.0.0-dev
libalire version:          3.0.0-dev
source commit:             unknown
source changes:            unknown
compilation date:          2025-04-16 09:18:48
compiled with version:     12.2.0

CONFIGURATION
home folder:               /home/jq
settings folder:           /home/jq/.config/alire
cache folder:              /home/jq/.local/share/alire
vault folder:              /home/jq/.local/share/alire/releases
build folder:              /home/jq/.local/share/alire/builds
temp folder:               /run/user/1000
force flag:                FALSE
non-interactive flag:      FALSE
community index branch:    stable-1.4.0
compatible index versions: ^1.1 & <=1.4.0
indexes folder:            /home/jq/.config/alire/indexes
indexes metadata:          OK
index #1:                  (community) git+https://github.com/alire-project/alire-index#stable-1.4.0
toolchain folder:          /home/jq/.local/share/alire/toolchains
toolchain assistant:       disabled
tool #1 gnat:              gnat_native=14.2.1
tool #2 gprbuild:          not configured
system package manager:    /usr/bin/apt
distro detection disabled: FALSE

WORKSPACE
root status:               VALID
root release:              hostarm=25.0.1
root load error:           none
root folder:               /home/jq/ada_examples/hostarm
current folder:            /home/jq/ada_examples/hostarm

SYSTEM
distribution:              DEBIAN
host-arch:                 X86_64
os:                        LINUX
target:                    NATIVE
toolchain:                 SYSTEM
word-size:                 BITS_64
```

**Output from alr**
```sh
...
Publishing assistant: step 6 of 11: Generate index manifest
✓ Your index manifest file has been generated at /home/jq/ada_examples/hostarm/alire/releases/hostarm-25.0.1.toml
Do you want to continue onto submission to the online community index?
[Y] Yes  [N] No  (default is Yes) 
Using default: Yes

Publishing assistant: step 7 of 11: Check existing PR
✓ No conflicting pull request found

Publishing assistant: step 8 of 11: Fork community index
✓ Community index fork exists in user account
ⓘ alr requires a GitHub Personal Access Token (PAT) to create a submission to the community index on your behalf. To avoid being asked for it every time, you can define the GH_TOKEN environment variable.
ⓘ You can create access tokens at https://github.com/settings/tokens
ⓘ Run alr help publish for further details.
Please provide your GitHub Personal Access Token:  (default: '')
> alr publish

Publishing assistant: step 9 of 11: Clone community index
✓ Local index updated successfully

Publishing assistant: step 10 of 11: Upload manifest
[release/hostarm-25.0.1 6ca96c1] hostarm 25.0.1 (via `alr publish`)
 1 file changed, 43 insertions(+)
 create mode 100644 index/ho/hostarm/hostarm-25.0.1.toml
fatal: unable to access 'https://github.com/jquorning/alire-index/': URL using bad/illegal format or missing URL
error: Could not complete the publishing assistant:
error:    Command ["git", "push", "-f", "-u", "writable", "release/hostarm-25.0.1"] exited with code 128
```

Ahh that did it :-)

Found some rough edges underway:

First I deleted the alire-index on github and got another error:
```sh
Publishing assistant: step 2 of 11: Verify GitHub infrastructure
✓ User has a GitHub account: jquorning
ⓘ alr requires a GitHub Personal Access Token (PAT) to fork the community index to your account. To avoid being asked for it every time, you can define the GH_TOKEN environment variable.
ⓘ You can create access tokens at https://github.com/settings/tokens
ⓘ Run alr help publish for further details.
Please provide your GitHub Personal Access Token:  (default: '')
> alr publish
A fork of the community index will now be created into your GitHub account to be able to submit a pull request. Do you agree?
[Y] Yes  [N] No  (default is Yes) 
Using default: Yes
error: Could not complete the publishing assistant:
error:    Attempt to fork repo [alire-index] owned by [alire-project] via GitHub REST API failed:
error:    Status line: HTTP/2 401:
error:    Response body:
error:    {:
error:    "message": "Bad credentials",:
error:    "documentation_url": "https://docs.github.com/rest",:
error:    "status": "401":
error:    }
```

But now it works! Thank You!

--- a/doc/publishing.md
+++ b/doc/publishing.md
@@ -31,6 +31,10 @@ and finally open a pull-request against the community repository.
 The PAT, once created, is a plain string. You can either export the environment
 variable `GH_TOKEN` set to this string, or provide it when Alire asks for it.
 
+**NOTE**: whenever Alire asks for the token, you need to provide the full
+string that Github shows only during its creation. Do not provide the token
+name instead.
+
 There are two kinds of PATs on GitHub: classic and fine-grained. The latter are
 in beta and not documented here yet. Follow these steps to create a classic PAT:
 

--- a/src/alire/alire-github.adb
+++ b/src/alire/alire-github.adb
@@ -5,8 +5,10 @@ with Alire.Errors;
 with Alire.OS_Lib;
 with Alire.Publish;
 with Alire.URI;
+with Alire.Utils.Regex;
 with Alire.Utils.TTY;
 with Alire.Version;
+with Alire.VCSs.Git;
 
 with GNATCOLL.JSON.Utility;
 with GNATCOLL.Strings;
@@ -510,5 +512,28 @@ package body Alire.GitHub is
            Outcome_Failure ("could not find release " & Tag, Report => False);
       end if;
    end Check_Alire_Binary_Release;
+
+   -------------------------
+   -- Is_Possibly_A_Token --
+   -------------------------
+
+   function Is_Possibly_A_Token (S : String) return Boolean
+   is
+   begin
+      --  https://docs.github.com/en/authentication/
+      --  keeping-your-account-and-data-secure/about-authentication-to-github
+      --  ?utm_source=chatgpt.com#githubs-token-formats
+
+      --  Classic PAT, just a 40 hexa string
+      if S in VCSs.Git.Git_Commit then
+         return True;
+      end if;
+
+      --  Newer PATs that start with gh?_ or github_pat_. Probably not all of
+      --  those can work, this would require further digging.
+      return Utils.Regex.Fully_Matches
+        (Regex => "(github_pat|gh[pours])_[a-f0-9]+",
+         Text  => S);
+   end Is_Possibly_A_Token;
 
 end Alire.GitHub;

--- a/src/alire/alire-github.ads
+++ b/src/alire/alire-github.ads
@@ -106,4 +106,8 @@ package Alire.GitHub is
    --  In case of error, the resulting Outcome will contain an error message
    --  specifying whether the release or the specific archive was not found.
 
+   function Is_Possibly_A_Token (S : String) return Boolean;
+   --  Check that S looks like a Personal Access Token. Not exhaustive, just
+   --  protect about obvious non-tokens
+
 end Alire.GitHub;

--- a/src/alire/alire-publish-submit.adb
+++ b/src/alire/alire-publish-submit.adb
@@ -54,6 +54,14 @@ package body Alire.Publish.Submit is
    -------------------
 
    function Ask_For_Token (Reason : String) return String is
+
+      ---------------
+      -- Non_Empty --
+      ---------------
+
+      function Non_Empty (S : String) return Boolean
+      is (S /= "");
+
       GH_Token : constant String := OS_Lib.Getenv (GitHub.Env_GH_Token, "");
    begin
       if GH_Token = "" then
@@ -79,7 +87,7 @@ package body Alire.Publish.Submit is
                  Validation =>
                  --  Allow bypassing in case some day GitHub introduces new
                  --  token formats so not to stop everybody completely.
-                   (if Force then null
+                   (if Force then Non_Empty'Unrestricted_Access
                     else GitHub.Is_Possibly_A_Token'Access)));
    end Ask_For_Token;
 

--- a/src/alire/alire-publish-submit.adb
+++ b/src/alire/alire-publish-submit.adb
@@ -54,14 +54,6 @@ package body Alire.Publish.Submit is
    -------------------
 
    function Ask_For_Token (Reason : String) return String is
-
-      --------------
-      -- Validate --
-      --------------
-
-      function Validate (S : String) return Boolean
-      is (S /= "");
-
       GH_Token : constant String := OS_Lib.Getenv (GitHub.Env_GH_Token, "");
    begin
       if GH_Token = "" then
@@ -81,9 +73,14 @@ package body Alire.Publish.Submit is
               then GH_Token
               else
                  CLIC.User_Input.Query_String
-                ("Please provide your GitHub Personal Access Token: ",
+                ("Please provide your GitHub Personal Access Token "
+                   & "(full value): ",
                  Default    => "",
-                 Validation => Validate'Unrestricted_Access));
+                 Validation =>
+                 --  Allow bypassing in case some day GitHub introduces new
+                 --  token formats so not to stop everybody completely.
+                   (if Force then null
+                    else GitHub.Is_Possibly_A_Token'Access)));
    end Ask_For_Token;
 
    ------------

--- a/src/alire/alire-utils-regex.adb
+++ b/src/alire/alire-utils-regex.adb
@@ -60,4 +60,29 @@ package body Alire.Utils.Regex is
 
       return "";
    end First_Match;
+
+   -------------
+   -- Matches --
+   -------------
+
+   function Matches (Regex : String; Text : String) return Boolean is
+   begin
+      return GNAT.Regpat.Match (Expression => Regex, Data => Text);
+   end Matches;
+
+   -------------------
+   -- Fully_Matches --
+   -------------------
+
+   function Fully_Matches (Regex : String; Text : String) return Boolean is
+   begin
+      if Regex = "" or else Regex (Regex'First) /= '^' then
+         return Fully_Matches ('^' & Regex, Text);
+      elsif Regex = "" or else Regex (Regex'Last) /= '$' then
+         return Fully_Matches (Regex & '$', Text);
+      else
+         return Matches (Regex, Text);
+      end if;
+   end Fully_Matches;
+
 end Alire.Utils.Regex;

--- a/src/alire/alire-utils-regex.ads
+++ b/src/alire/alire-utils-regex.ads
@@ -12,4 +12,12 @@ package Alire.Utils.Regex is
    --  is the second matching expression. In case of no match, it will return
    --  an empty string. At least one capture must be attempted in the Regex.
 
+   function Matches (Regex : String; Text : String) return Boolean;
+   --  Wrapper on GNAT.Regpat, just say if Text matches Regex. Remember to use
+   --  anchors (^$) for full string matching, as otherwise a substring match is
+   --  reported as a valid match. See Fully_Matches below.
+
+   function Fully_Matches (Regex : String; Text : String) return Boolean;
+   --  Returns true if Text is matched in full by regex
+
 end Alire.Utils.Regex;

--- a/src/alire/alire-utils.ads
+++ b/src/alire/alire-utils.ads
@@ -8,6 +8,10 @@ package Alire.Utils with Preelaborate is
    subtype Hexadecimal_Character is Character with
      Static_Predicate => Hexadecimal_Character in '0' .. '9' | 'a' .. 'f';
 
+   subtype Hexadecimal_String is String with
+     Predicate => (for all Char of Hexadecimal_String =>
+                     Char in Hexadecimal_Character);
+
    function Command_Line_Contains (Prefix : String) return Boolean;
    --  Say if any of the command-line arguments begins with Prefix. This is
    --  needed for string arguments, that even when not supplied are initialized

--- a/testsuite/tests_ada/src/alr_tests-github_tokens.adb
+++ b/testsuite/tests_ada/src/alr_tests-github_tokens.adb
@@ -1,0 +1,14 @@
+with Alire.GitHub;
+
+procedure Alr_Tests.Github_Tokens is
+   use Alire.GitHub;
+begin
+   pragma Assert (Is_Possibly_A_Token ("12345678abcdefab12345678abcdefab12345678"),
+                  "Failed for old GitHub PAT format");
+   pragma Assert (Is_Possibly_A_Token ("github_pat_1234567890abcdef"),
+                  "Failed for GitHub fine-grained PAT format");
+   pragma Assert (Is_Possibly_A_Token ("ghp_1234567890abcdef"),
+                  "Failed new GitHub PAT format");
+   pragma Assert (not Is_Possibly_A_Token ("MY_PERSONAL_TOKEN"),
+                  "Should not accept invalid GitHub PAT format");
+end Alr_Tests.Github_Tokens;

--- a/testsuite/tests_ada/src/alr_tests-regexes.adb
+++ b/testsuite/tests_ada/src/alr_tests-regexes.adb
@@ -1,0 +1,37 @@
+with Alire.Utils.Regex;
+
+procedure Alr_Tests.Regexes is
+   use Alire.Utils.Regex;
+
+   GH_Token_Regex : constant String := "(github_pat|gh[pours])_[a-f0-9]+";
+
+begin
+   --  Partial matches with unanchored expressions
+   pragma Assert (Matches ("abc", "abcde"), "Failed unanchored match at beginning");
+   pragma Assert (Matches ("cde", "abcde"), "Failed unanchored match at end");
+   pragma Assert (Matches ("bcd", "abcde"), "Failed unanchored match in the middle");
+
+   --  Full matches with unanchored expressions
+   pragma Assert (Fully_Matches ("abc", "abc"), "Failed full match");
+   pragma Assert (not Fully_Matches ("abc", "abcde"), "Should not partially match");
+
+   --  Anchored full matches
+   pragma Assert (Matches ("^abc$", "abc"), "Failed full anchored match");
+   pragma Assert (Fully_Matches ("^abc$", "abc"), "Failed redundant full anchored match");
+
+   --  Some failing matches
+   pragma Assert (not Matches ("abc", "def"), "Should not match different strings");
+   pragma Assert (not Matches ("abc", "ab"), "Should not match shorter string");
+   pragma Assert (not Fully_Matches ("abc", "abcd"), "Should not match longer string");
+
+   --  Some actual regex tests
+   pragma Assert (Matches ("^\d{3}-\d{2}-\d{4}$", "123-45-6789"), "Failed regex match for SSN format");
+   pragma Assert (Fully_Matches ("[0-9a-f]{8}", "1a2b3c4d"), "Failed regex match for hex format");
+   pragma Assert (not Fully_Matches ("[0-9a-f]{8}", "1a2b3c4d5"), "Should not match longer hex format");
+   pragma Assert (Fully_Matches (GH_Token_Regex, "github_pat_1234567890abcdef"),
+                  "Failed regex match for GitHub fine-grained PAT format");
+   pragma Assert (Fully_Matches (GH_Token_Regex, "ghp_1234567890abcdef"),
+                  "Failed match for new GitHub PAT format");
+   pragma Assert (not Fully_Matches (GH_Token_Regex, "MY_PERSONAL_TOKEN"),
+                  "Should not match invalid GitHub PAT format");
+end Alr_Tests.Regexes;

